unified_planning/model/htn/method.py

```diff
@@ -0,0 +1,239 @@
+    """This is the method interface."""
+    def __init__(self, _name: str, _parameters: 'OrderedDict[str, up.model.types.Type]' = None,
+                 _env: Environment = None, **kwargs: 'up.model.types.Type'):
+        self._env = get_env(_env)
+        self._task: Optional[ParameterizedTask] = None
+        self._name = _name
+        self._parameters: 'OrderedDict[str, Parameter]' = OrderedDict()
+        self._preconditions: List[up.model.fnode.FNode] = []
+        self._constraints: List[up.model.fnode.FNode] = []
+        self._subtasks: List[Subtask] = []
+        if _parameters is not None:
+            assert len(kwargs) == 0
+            for n, t in _parameters.items():
+                self._parameters[n] = Parameter(n, t)
+        else:
+            for n, t in kwargs.items():
+                self._parameters[n] = Parameter(n, t)
+
+    def __repr__(self) -> str:
+        s = []
+        s.append(f'method {self.name}')
+        first = True
+        for p in self.parameters:
+            if first:
+                s.append('(')
+                first = False
+            else:
+                s.append(', ')
+            s.append(str(p))
+        if not first:
+            s.append(')')
+        s.append(' {\n')
+        s.append(f'  task = {self._task}\n')
+        s.append('  preconditions = [\n')
+        for c in self.preconditions:
+            s.append(f'    {str(c)}\n')
+        s.append('  ]\n')
+        s.append('  constraints = [\n')
+        for c in self.constraints:
+            s.append(f'    {str(c)}\n')
+        s.append('  ]\n')
+        s.append('  subtasks = [\n')
+        for st in self.subtasks:
+            s.append(f'      {str(st)}\n')
+        s.append('  ]\n')
+        s.append('}')
+        return ''.join(s)
+
+    def __eq__(self, oth: object) -> bool:
+        if not isinstance(oth, Method):
+            return False
+        return (self._env == oth._env and
+                self._name == oth._name and
+                self._parameters == oth._parameters and
+                self._task == oth._task and
+                set(self._preconditions) == set(oth._preconditions) and
+                set(self._subtasks) == set(oth._subtasks) and
+                set(self._constraints) == set(oth._constraints))
+
+    def __hash__(self) -> int:
+        res = hash(self._name)
+        res += hash(self._task)
+        res += sum(map(hash, self.parameters))
+        res += sum(map(hash, self._preconditions))
+        res += sum(map(hash, self._constraints))
+        res += sum(map(hash, self.subtasks))
+        return res
+
+    @property
+    def name(self) -> str:
+        """Returns the action name."""
+        return self._name
+
+    @property
+    def achieved_task(self) -> ParameterizedTask:
+        """Returns the task that this method achieves."""
+        assert self._task is not None
+        return self._task
+
+    def set_task(self, task: Union[Task, ParameterizedTask], *arguments: Parameter):
```

Comment:
In this methods (and other places) asserts are ued to report errors that can arise by API misuse. I would use a `ValueError` for these cases (that is not disabled by `python -O`)

Reply:
Hummm, I had very similar discussion with @Framba-Luca a few days ago where I essentially declared my love of asserts : https://github.com/aiplan4eu/unified-planning/pull/141#issuecomment-1124926482

In particular here, the checks have some associated computational cost (e.g. membership of a list, type checking) so one might want to disable them if they are sure that they use the API correctly. Those are essentially programming error for which we should not try to recover but we just want to warn the user so that it does not shoot himself in the foot.

I do use specialized exception in the PDDL reader where we should be able to recover from an invalid input (and in fact I converted an assert to a `SyntaxError` after double checking.

Reply:
We disagree on this one :smiling_imp: 

In my opinion asserts should only be used for code invariants, something that your code guaranteed to hold. Validation of user inputs should raise a proper exception, especially in this library where we aim to make it as user-friendly as possible. (More or less I agree with this blog: https://stefan.sofa-rockers.org/2018/04/16/assertions-and-exceptions/)

Not a big deal for this round, though, we can merge and fix this later on.

Reply:
Ok, I guess I will play by the local rules then =)
I'll try to revisit those asserts incrementally next time I am on this part of the code.